components/org.wso2.carbon.appmgt.gateway/src/main/java/org/wso2/carbon/appmgt/gateway/handlers/proxy/ReverseProxyHandler.java

```diff
@@ -115,4 +109,88 @@ public boolean handleResponse(MessageContext messageContext) {
 
 		return true;
 	}
+
+	/**
+	 * Fixes the cookie paths if there exists any in the response so that it matches only to this context.
+	 * Otherwise cookies destined to other applications will also arrive onto this application. Because most cookies are
+	 * set to the root context or the context of the the original web application.
+	 *
+	 * @param axis2MC
+	 * @param headers
+	 * @param webContextWithVersion
+	 */
+	private void fixCookiePaths(org.apache.axis2.context.MessageContext axis2MC, TreeMap headers,
+			String webContextWithVersion) {
+		Map<String, Object> excessHeaders = (Map<String, Object>) axis2MC.getProperty(NhttpConstants.EXCESS_TRANSPORT_HEADERS);
+
+		List<String> excessCookies = (List<String>) excessHeaders.get(HTTPConstants.HEADER_SET_COOKIE);
+
+		if(excessCookies != null) {
+			List<String> fixedCookies = new ArrayList<>();
+			for (String cookie : excessCookies) {
+				List<HttpCookie> httpCookies = HttpCookie.parse(cookie);
+				for (HttpCookie httpCookie : httpCookies) {
+					String newPath = replaceCookieContextPath(webContextWithVersion, httpCookie);
+					httpCookie.setPath(newPath);
+					fixedCookies.add(toHeaderString(httpCookie));
+				}
+			}
+
+			if (fixedCookies.size() > 0) {
+				excessHeaders.remove(HTTPConstants.HEADER_SET_COOKIE);
+				for (int i = 0; i < fixedCookies.size(); i++) {
+					excessHeaders.put(HTTPConstants.HEADER_SET_COOKIE, fixedCookies.get(i));
+				}
+			}
+		}
+
+		Object cookieFromHeader = headers.get(HTTPConstants.HEADER_SET_COOKIE);
```

Comment:
if(cookieFromHeader == null) then we do not need to fix the cookies in EXCESS_TRANSPORT_HEADERS